### ISSUE TITLE
Binary ops

### ISSUE THREAD

**Issue by bhack** [tensorflow/tensorflow#1592]
Is there already a plan to add binary ops like bitcount for [XNOR-NET](http://arxiv.org/abs/1603.05279)?

**Comment by vrv** [comment]
I think this would be awesome to have, and contributions are definitely welcome here :).

**Comment by bhack** [comment]
@vrv Probably we need a XnorGemm first in Eigen. /cc @benoitsteiner What do you think?

**Comment by vrv** [comment]
Or we could also implement them as individual OpKernels if it is too difficult to get it into Eigen.

**Comment by bhack** [comment]
This is a reference XnorGemm (with a custom kernel) under BSD related to a previous paper: https://github.com/MatthieuCourbariaux/BinaryNet/blob/master/Run-time/binary_ops.py

**Comment by bhack** [comment]
@scott-gray is working on an [improved version with the upstream author](https://github.com/soumith/convnet-benchmarks/issues/93#issuecomment-200518733). Scott will you release the code under BSD or Apache? Eigen library is currently on BSD and TF on Apache but needs cla signature.

**Comment by bhack** [comment]
@benoitsteiner Do you think that this operations could be added in Eigen first?

**Comment by benoitsteiner** [comment]
@bhack we have a set of Eigen extensions to better support quantized operations on tensors in https://github.com/tensorflow/tensorflow/tree/master/third_party/eigen3/unsupported/Eigen/CXX11/src/FixedPoint. It's definitely possible to use the same approach to package a XnorGemm operation.
I can also talk to the other maintainers of Eigen to check if it makes sense to add the code into the core Eigen and make it more widely available.

**Comment by bhack** [comment]
@benoitsteiner Yes could be useful if you can collect some upstream opinions.

**Comment by bhack** [comment]
8 bit quantization seems available now. See merged https://github.com/tensorflow/tensorflow/pull/2230

**Comment by kofd** [comment]
Has there been any progress on this? It would be useful for some embedded applications where an nvidia gpu isn't an option

**Comment by bhack** [comment]
@kofd You can also start to read https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/how_tos/quantization/index.md

**Comment by petewarden** [comment]
I have been looking at 'popcount' for binary networks, as bitcount is often known, since that seems to be the trickiest part to map to processor instructions. There is some BSD-licensed work here:
https://github.com/WojciechMula/sse-popcount
Interestingly the x86 CPU instruction seems to be competitive with SSE implementations. It looks like ARM requires a multi-instruction macro though:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0081b/CHDJJGAJ.html

**Comment by bhack** [comment]
@peterwarden There are also GCC built-in and llvm intrinsic. How many compilers TF want support?

**Comment by kofd** [comment]
@bhack: I was talking about the bit count convolutions used in xnor net.

**Comment by bhack** [comment]
@peterwarden A simple test of built-in with GCC and msvc at https://github.com/hanji/popcnt/blob/master/populationcount.cpp I think it is easy to add also llvm intrinsic for popcount.

**Comment by zhengwy888** [comment]
I have implemented a primitive op on cpu for the XOR + bitcount. but it's too slow right now. Does any one know how to speed this up? if this ever gets to the same speed as tf.matmul than I should provide a patch.

```c++
  void concatenate_col(
          typename MatMulTypes<T>::in_type array,
          MaskMatrix &out)
  {
      int rowSize = int((array.dimension(0)+31)/ 32);
      out.resize(array.dimension(1),rowSize );

      for ( int c=0; c< array.dimension(1); ++ c)
      {
          for ( int r = 0; r < rowSize; ++ r )
          {
              uint32_t rvalue=0;
              uint32_t sign;
              for ( int i=0; i< 32; ++i ) {
                  int rowIdx = r*32 + i;
                  if ( rowIdx > array.dimension(0)-1 ) {
                      break;
                  }
                  sign = (array(rowIdx, c )>=0);
                  rvalue = rvalue | (sign <<i);
              }
              out(c, r) = rvalue;
          }
      }
  }
  void concatenate_row(
          typename MatMulTypes<T>::in_type array,
          MaskMatrix &out)
  {
      int colSize = int((array.dimension(1)+31 )/ 32);
      out.resize(array.dimension(0), colSize);
      for ( int r = 0; r < array.dimension(0); ++ r )
      {
          for ( int c=0; c< colSize; ++ c)
          {
              uint32_t rvalue=0;
              uint32_t sign;
              for ( int i=0; i< 32; ++i ) {
                  int colIdx = c*32 + i;
                  if ( colIdx > array.dimension(1)-1 ) {
                      break;
                  }
                  sign = (array(r, colIdx)>=0);
                  rvalue = rvalue | (sign <<i);
              }
              out(r,c) = rvalue;
          }
      }
  }

  void concatenate_and_compute(
          const CPUDevice &d,
          typename MatMulTypes<T>::in_type a,
          typename MatMulTypes<T>::in_type b,
          typename MatMulTypes<T>::out_type out)
  {
      MaskMatrix a_;
      MaskMatrix b_;

      concatenate_row(a, a_);
      concatenate_col(b, b_);

      for (int ar=0; ar < a_.rows(); ar++)
      {
          for (int br=0; br< b_.rows(); br++) {
              unsigned int Cvalue = 0;
              for (int c=0; c< a_.cols(); c++)
              {
                  unsigned int value =popcnt(a_(ar, c) ^ b_(br,c));
                  Cvalue += value;
              }
              out(ar, br) = - ( 2*(float)Cvalue - a.dimension(1) );
          }
      }

  }

```

**Comment by ppwwyyxx** [comment]
I've worked on binary GEMM/convolution before. From my experience the best approach for popcnt on avx2 would be this one: https://github.com/WojciechMula/sse-popcount/blob/master/popcnt-avx2-lookup.cpp. But that code needs a little fix for a counter overflow.  xor also needs to be done in avx2.
For TF I guess there needs to be a generic non-avx2 code. There are some references in that repo as well.

**Comment by bhack** [comment]
@ppwwyyxx Have you benchmarked  against recent GCC, MSVC, LLVM/CLANG intrinsic?

**Comment by ppwwyyxx** [comment]
@bhack I don't think there will be a big difference with different intrinsics. They all end up being avx2 instructions anyway.

**Comment by bhack** [comment]
Are you sure that all internal compiler code use avx2? I think that built-in are supported also on arm/neon.

**Comment by ppwwyyxx** [comment]
Oh right. if you are talking about compatibility then compiler builtins may be a good choice. But i never tried them.

**Comment by isabel-schwende** [comment]
Has there been any movement in this issue? I'm very interested in seeing how binary networks can be trained using tensorFlow. I have studied the work of Courbariaux and played a bit with his implementations (specifically BinaryConnect) but my final goal would be to have XNOR-Net running with tensorFlow.

**Comment by lorenlugosch** [comment]
would also be interested in this!

**Comment by zhengwy888** [comment]
binarization can be done with tf.sign(). though the tricky part is to get the gradient backprop to work after binarizing input. for now this requires a separate op I implemented in tensorflow. https://github.com/zhengwy888/binary_ops. with this code you can implement your own XNOR net on GPU. comments/suggestions welcome.

**Comment by bhack** [comment]
For who is interested there is also https://arxiv.org/abs/1606.06160

**Comment by ppwwyyxx** [comment]
Thanks @bhack for mentioning. We have a DoReFa-Net training implementation available [here](https://github.com/ppwwyyxx/tensorpack/tree/master/examples/DoReFa-Net), which doesn't make use of any custom Op. Since DoReFa-Net is a generalization of XNOR-Net, XNOR-Net can be built in TF in a similar manner, if it is only for training purpose. I'm also releasing a trainable DoReLa(1,2,6)-AlexNet later today.

**Comment by bhack** [comment]
@ppwwyyxx Nice! /cc @wangyida

**Comment by isabel-schwende** [comment]
Thanks for sharing everyone. But as I understood the DoReFa implementation so far, it still uses the standard 32 bit tensors. I've been thinking about ways to use the official TensorFlow quantization methods to reduce the memory at least to a quarter. The 8-bit datatype is available. Of course the question would be how to use it with the least hassle.

**Comment by Wangyida** [comment]
In my perspective of view. released pipeline of DeRoFa is still float, and quantization module in TF could have a 8 bit representation with little performance drop, but this won't be the aim of DeRoFa which is already a quantized model just in a float representation of codes.

**Comment by isabel-schwende** [comment]
@Wangyida Yes, I agree that it seems that the intention of the authors of DoReFa was mostly to reduce training and inference time by using low bitwidth. I don't expect them to release an 8bit version. However they also mention in their paper the idea to use this kind of network on embedded devices. If you want to use AlexNet on a very small device, I don't see the reason to use 32 bit floats if there is no information held in them anyways. To me, the quantization to a smaller datatype is just the next logical step.

**Comment by ppwwyyxx** [comment]
@isabel-schwende Yes the released implementation uses float32 to __hold__ low-bitwidth numbers, because there is simply no low-bitwidth operations available in TF. And we never planned to build such operations into TF because we already have our own run-time implementations working smoothly on ARM.  
The released model is similar: it uses tf.float32 to hold all the binary weights as well as run all the computation. But anyone who would like to implement those binary operations can make use of our pretrained model directly.

**Comment by ppwwyyxx** [comment]
I totally agree and the code we released was never intended for running on mobiles, but for showing how to train such networks in tensorflow, as a supplementary material and proof of the paper.

I can certainly compress our models to about 30x smaller because they are essentially binary, and maybe use tf.quint8 for computation. But that doesn't make sense to me, because I think anyone who really want 1 or 2 bit level of performance & compression rate, should use a much more compact & tiny run-time toolchain as we did, instead of using tensorflow. If the speed & storage of 8 bit models is good enough for the use case, then I would certainly suggest trying the TF quantization tutorial instead of DoReFa-Net, because 8 bit models would have better accuracy.

**Comment by isabel-schwende** [comment]
@ppwwyyxx Thanks a lot for your answer. I definitely agree. For that reason I started with the quantization available by TensorFlow. It seems like using it for inference is currently very slow. As several others have described here https://github.com/tensorflow/tensorflow/issues/2807 the memory might be much lower for saving the model but inference time is approximately doubled, which is undesirable in our project. I've been searching for something that is fast and has low memory requirements but it seems like the available code is still in research phase. I guess we will do some hacking to get something customised for our project.

**Comment by petewarden** [comment]
The current slow speed of quantization is due to reference implementations for some of the ops, which we're actively working on optimizing. The goal is that quantization should be faster than float.

**Comment by AaronYKing** [comment]
Hi All,
Is it implemented through code about XNOR-Net? If it can be implemented by Caffe?

**Comment by bhack** [comment]
See also https://github.com/NervanaSystems/neon/commit/caf0aaaaa1438b09c905e0780ba1120c6fd25f1c

**Comment by AaronYKing** [comment]
@bhack It is binary net not XNOR-Net! Anyway thank you for the info.

**Comment by isabel-schwende** [comment]
@AaronYKing I guess certain companies might have their own working versions of XNOR-Net but, as far as I know, there is no complete public implementation so far. It's too bad that the original authors didn't release their code for XNOR-Net for the sake for reproducibility. If I remember correctly, the authors of DoReFa-Net tried to reproduce results of the original paper but failed to obtain comparable results. As the description in the paper is a bit thin at times, it might take a while for a really comparable, complete implementation of XNOR to be released (no matter if TensorFlow or Caffe).

**Comment by mrastegari** [comment]
We have released our Torch code and trained models for XNOR-Net. This is not the fast implementation.

**Comment by mrastegari** [comment]
Here is the link: https://github.com/mrastegari/XNOR-Net

**Comment by albicasty** [comment]
@mrastegari  thank you. Do you have a version with the proper XNOR convolution and not using the SpatialConvolution. I am  interested in the backward pass of  BinConv.

**Comment by AaronYKing** [comment]
@isabel-schwende Thank you very much for your patient replay. Now the author have released the torch code.

**Comment by AaronYKing** [comment]
@mrastegari Thank you for your contribution. I wonder if it can be implemented by Caffe and will anybody implement it in the near future？

**Comment by isabel-schwende** [comment]
@mrastegari thanks a lot for sharing this information with us. I think this is going to be really helpful. Do you have any numerical information of how slow this Torch version is compared to the original version using Darknet? @AaronYKing I'm always glad if there are cases where I'm positively surprised about code being shared publicly.

**Comment by bhack** [comment]
Is this xnor gemm kernel useful https://github.com/NervanaSystems/neon/blob/master/neon/backends/kernels/cuda/binary.py?

**Comment by ppwwyyxx** [comment]
@bhack This kernel works, but only about 3.4x faster than best fp32 kernels (cublas). It's mentioned in the BNN paper and https://github.com/MatthieuCourbariaux/BinaryNet/pull/1.

**Comment by dmitryb-ai2** [comment]
Just curious, is anyone seriously pursuing this? I've been working on a fast C++ implementation of XNORNets here at AI2 with @mrastegari and others for Intel and ARM CPUs. We've achieved a modicum of success, with much headroom still remaining untapped. We're kicking around the idea of producing a fast, reference CPU implementation, so it'd be good to know if someone else is already close to releasing it.

**Comment by csyking** [comment]
@dmitryb-ai2 @mrastegari It will be a great progress if the XNORNets can be implemented by a fast C++ for Intel  and ARM CPUs, especially getting 58× faster convolutional operations and 32× memory savings as mentioned in the paper. But, sadly, I can't code it. So looking forward to your good news!

**Comment by lorenlugosch** [comment]
I have an "implementation", but it achieves only ~15% accuracy on CIFAR-10
(compared to 86% using real-valued convolutions for the same architecture),
so something is definitely wrong. I will post my code if and when I get it
to work.

**Comment by csyking** [comment]
@lorenlugosch Hi, I am interested in your problem. What framework do you use? And will you release it to the git?

**Comment by csyking** [comment]
Hi @lorenlugosch, It's very kind of you. Looking forward to your good new!

**Comment by albicasty** [comment]
@mrastegari  May you share the model for the Resnet-14? I am training it starting from the model you proposed in your implementation (with some modifications) but the progresses seem to be very slow.

**Comment by rapatel0** [comment]
We got a "version" working in Eigen::Tensor (~7x performance over float on Xeon AVX-256), but we're still hitting relatively low accuracies (20 to 40 % more error than a float based net). The accuracy drops quickly as you increase the output channels in a layer.  From a performance POV the bit packing code slows things down a bit and the calculation of beta values still takes time. 
I'm not sure if the original paper took this into account because it defined a "convolution" as seperate from the binarize step. In practice, however, you need both for every conv2d layer. Still, I might be missing something.

**Comment by zhengwy888** [comment]
@rapatel0 7x performance means 7x faster? could you release some part of your code? I tried to use gcc popcnt for bitcounting but it's really slow.

**Comment by rapatel0** [comment]
Depending on your compiler flags gcc popcount won't emit a _popcnt64 instruction which limits the performance. The code is still buggy and requires some more testing but I'll release it once we get it cleaned up.

**Comment by bhack** [comment]
What do you think of [this Adobe patent](http://www.freepatentsonline.com/y2016/0148078.html)?

**Comment by bhack** [comment]
For who is still interested in the topic see QNN https://arxiv.org/abs/1609.07061

**Comment by eamartin** [comment]
Any progress on this? Besides more sophisticated kernels like XnorGEMM, there's value in supporting bitwise and, or, xor on types like int32 and int64.

**Comment by Randl** [comment]
Definitely very low precision networks become more and more popular: see [Trained Ternary Quantization](https://arxiv.org/abs/1612.01064) and [Incremental Network Quantization: Towards Lossless CNNs with Low-Precision Weights](https://arxiv.org/abs/1702.03044). We are looking forward for arbitrary quantization and efficient operations (e.g. XNOR-popcount) in TensorFlow

**Comment by bhack** [comment]
Or in gemmlowp so other frameworks could use it.

**Comment by Alexivia** [comment]
@lorenlugosch Did you managed to finish your binary convolution? I'm interested in knowing how to implement low-precision operations for the forward pass in CNNs... @bhack Do you know if there is any support for this in TF? Besides the 8bit quantisation process, I want a "hard" quantisation, not the one that uses the max and min with float values

**Comment by Cogitans** [comment]
Does anyone have an update on progress towards binary TF ops? I'm weighing the pros and cons of working on this problem myself (the pro being that it'll be useful, and the con primarily being the technical investment of fully grokking the Eigen/cuBLAS side of TF).

Thanks!

**Comment by Randl** [comment]
https://github.com/Microsoft/CNTK/tree/master/Examples/Extensibility/BinaryConvolution
https://github.com/hpi-xnor/BMXNet

Binary nets in other frameworks

**Comment by AngusG** [comment]
It's still a bit rough, but I thought i'd share my custom op with an mnist training example and benchmarks against tf.matmul. Feedback and suggestions welcome!

https://github.com/AngusG/tensorflow-xnor-bnn

**Comment by abhishek42** [comment]
#https://github.com/MatthieuCourbariaux/BinaryNet/issues/21

**Comment by annarev** [comment]
ebrevdo@, I found that you added PopulationCount op some time ago:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/bitwise_ops.cc
which seems like it supports 'bitcount':
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/base_api/api_def_PopulationCount.pbtxt

Is this op available to use or is there more work needed?

**Comment by ebrevdo** [comment]
The op is supported for use; but not part of the public API; you can access it directly though.  See an example [here](https://github.com/tensorflow/tensorflow/blob/40d1b257636a0b510dba59aff0af54e42d602313/tensorflow/python/ops/bitwise_ops_test.py#L69)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
Marking as closed since it's available.

**Comment by mohendra** [comment]
gemm_op.so not find

**Comment by arashb** [comment]
> Is there already a plan to add binary ops like bitcount for [XNOR-NET](http://arxiv.org/abs/1603.05279)?

If someone is still interested, we have developed open source libraries on top of @tensorflow and TensorFlow Lite to train and deploy binarized neural networks (BNNs) similar to XNOR-NET.
- [Larq](https://github.com/larq/larq) to train BNNs.
- [Larq Compute Engine](https://github.com/larq/compute-engine ) to deploy BNNs on Arm devices.
